module/state_synchronization/indexer/persister.go

```diff
@@ -0,0 +1,239 @@
+package indexer
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/store/inmemory/unsynchronized"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// BlockPersister handles transferring data from in-memory storages to permanent storages
+// for a single block.
+//
+// - Each BlockPersister instance is created for ONE specific block
+// - All `inMemory*` storages contain data ONLY for this specific block, they are not shared
+type BlockPersister struct {
+	log zerolog.Logger
+
+	inMemoryRegisters      *unsynchronized.Registers
+	inMemoryEvents         *unsynchronized.Events
+	inMemoryCollections    *unsynchronized.Collections
+	inMemoryTransactions   *unsynchronized.Transactions
+	inMemoryResults        *unsynchronized.LightTransactionResults
+	inMemoryTxResultErrMsg *unsynchronized.TransactionResultErrorMessages
+
+	registers      storage.RegisterIndex
+	events         storage.Events
+	collections    storage.Collections
+	transactions   storage.Transactions
+	results        storage.LightTransactionResults
+	txResultErrMsg storage.TransactionResultErrorMessages
+	protocolDB     storage.DB
+
+	executionResult *flow.ExecutionResult
+	header          *flow.Header
+}
+
+// NewPersister creates a new persister.
+func NewPersister(
+	log zerolog.Logger,
+	inMemoryRegisters *unsynchronized.Registers,
+	inMemoryEvents *unsynchronized.Events,
```

Comment:
I wonder if makes sense to take a RegisterPersister instance instead here.

The RegisterPersister can take `*unsynchronized.Events ` and `registers storage.Events ` (maybe also header and excutionResult) and provide a `Persist(batch)` method to do `addEventsToBatch`.

We can also abstract it into PersisterStore interface which provides the `Persist(batch)` method, and then this NewPersister can just take a list of `PersisterStore`. 

Registers might have to be treated separately, because it's stored in a separate database.